src/RecordTable.h

```diff
@@ -0,0 +1,158 @@
+
+#pragma once
+
+#include "CompiledTuple.h"
+#include "ParallelUtils.h"
+#include "RamTypes.h"
+#include <cassert>
+#include <iostream>
+#include <limits>
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+namespace souffle {
+
+/**
+ * A bidirectional mapping between tuples and reference indices.
+ */
+class RecordMap {
+    /** The arity of the stored tuples */
+    int arity;
+
+    /** The mapping from tuples to references/indices */
+    std::map<std::vector<RamDomain>, RamDomain> r2i;
+
+    /** The mapping from indices to tuples */
+    std::vector<std::vector<RamDomain>> i2r;
+
+public:
+    RecordMap(int arity) : arity(arity), i2r(1) {}  // note: index 0 element left free
+
+    /**
+     * Packs the given tuple -- and may create a new reference if necessary.
+     */
+    RamDomain pack(const RamDomain* tuple) {
+        std::vector<RamDomain> tmp(arity);
+        for (int i = 0; i < arity; i++) {
+            tmp[i] = tuple[i];
+        }
+
+        RamDomain index;
+#pragma omp critical(record_pack)
+        {
+            auto pos = r2i.find(tmp);
+            if (pos != r2i.end()) {
+                index = pos->second;
+            } else {
+#pragma omp critical(record_unpack)
+                {
+                    i2r.push_back(tmp);
+                    index = i2r.size() - 1;
+                    r2i[tmp] = index;
+
+                    // assert that new index is smaller than the range
+                    assert(index != std::numeric_limits<RamDomain>::max());
+                }
+            }
+        }
+
+        return index;
+    }
+
+    /**
+     * Obtains a pointer to the tuple addressed by the given index.
+     */
+    RamDomain* unpack(RamDomain index) {
+        RamDomain* res;
+
+#pragma omp critical(record_unpack)
+        res = &(i2r[index][0]);
+
+        return res;
+    }
+};
+
+class RecordTable {
+public:
+    RecordTable() = default;
+    virtual ~RecordTable() = default;
+
+    /**
+     * A function packing a tuple of the given arity into a reference.
+     */
+    RamDomain pack(RamDomain* tuple, int arity) {
+        return getForArity(arity).pack(tuple);
+    }
+
+    /**
+     * A function packing a tuple of the given arity into a reference.
+     */
+    template <typename Domain, std::size_t _arity>
+    RamDomain pack(ram::Tuple<Domain, _arity> tuple) {
+        static_assert(sizeof(Domain) == sizeof(RamDomain),
+                "Pointer cast: Size of tuple domain must equal domain in Record Table.");
+        return getForArity(_arity).pack(static_cast<RamDomain*>(tuple.data));
+    }
+
+    /**
+     * A function obtaining a pointer to the tuple addressed by the given reference.
+     */
+    RamDomain* unpack(RamDomain ref, int arity) {
+        return getForArity(arity).unpack(ref);
+    }
+
+    /**
+     * A function obtaining a pointer to the tuple addressed by the given reference.
+     */
+    template <typename Domain, std::size_t _arity>
```

Comment:
You could use `Arity` instead of `_arity`, which would be more in keeping with our usual naming strategies.